R/pivot-wide.R

```diff
@@ -542,16 +546,44 @@ select_wider_id_cols <- function(data,
                                  non_id_cols = character()) {
   id_cols <- enquo(id_cols)
 
-  # Remove known non-id-cols so they are never selected
+  # Remove known `non_id_cols` so they are never selected
   data <- data[setdiff(names(data), non_id_cols)]
```

Comment:
Are you sure it wouldn't be better to leave data as is and then error if any of `non_id_cols` appear in `id_cols`?

Reply:
I wanted to ensure that using something like `id_cols = everything()` would exactly match the default `id_cols = NULL` behavior, i.e.:

``` r
library(tidyr)

df <- tibble(
  id = c("a", "a", "b"),
  name = c("x", "y", "x"), 
  value = c(1, 2, 2)
)
df
#> # A tibble: 3 × 3
#>   id    name  value
#>   <chr> <chr> <dbl>
#> 1 a     x         1
#> 2 a     y         2
#> 3 b     x         2

pivot_wider(
  df,
  id_cols = everything(), # <- exactly matches default `NULL` behavior. "everything" except names_from/values_from
  names_from = name,
  values_from = value
)
#> # A tibble: 2 × 3
#>   id        x     y
#>   <chr> <dbl> <dbl>
#> 1 a         1     2
#> 2 b         2    NA
```

This works in both tidyr 1.2.0 _and_ 1.1.4. It works in 1.1.4 because we called `eval_select()` on all of `data` as you are saying here, and then just did a `setdiff()` on the `non_id_cols`. But this resulted in the original confusing case where you could specify a particular column in both `id_cols` and `names/values_from` and it would get silently (and confusingly) removed from `id_cols`.

If we left `data` as is and errored on any `id_cols` that were also in `non_id_cols`, then we'd always error on the `id_cols = everything()` case, because it would always select something that overlaps with both `names_from` and `values_from`. Since this used to work in pre 1.2.0 tidyr, I'd be worried that it may break someone's code.

You could also imagine some usage of `starts_with()` or any other tidyselect helper that is dynamic being used for `id_cols`, and you could run into the same issue.

Reply:
Got it!